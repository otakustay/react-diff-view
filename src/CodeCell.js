import {PureComponent} from 'react';

export default class CodeCell extends PureComponent {

    static defaultProps = {
        onRender() {
        }
    };

    cell = null;

    componentDidMount() {
        this.notifyRender();
    }

    componentDidUpdate(prevProps) {
        if (prevProps.text !== this.props.text || prevProps.html !== this.props.html) {
            this.notifyRender();
        }
    }

    notifyRender() {
        this.props.onRender(this.cell);
    }

    render() {
        /* eslint-disable no-unused-vars */
        const {text, html, onRender, ...props} = this.props;
        /* eslint-enable no-unused-vars */
        const ref = cell => (this.cell = cell);

        if (html) {
            /* eslint-disable react/no-danger */
            return <td ref={ref} {...props} dangerouslySetInnerHTML={{__html: html}} />;
            /* eslint-enable react/no-danger */
        }

        return <td ref={ref} {...props}>{text}</td>;
    }
}
