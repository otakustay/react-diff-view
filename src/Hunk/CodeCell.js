import {memo} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const defaultRenderToken = ({type, value, markType, properties, children}, i) => {
    const renderWithClassName = className => (
        <span key={i} className={className}>
            {value ? value : (children && children.map(defaultRenderToken))}
        </span>
    );

    switch (type) {
        case 'text':
            return value;
        case 'mark':
            return renderWithClassName(`diff-code-mark diff-code-mark-${markType}`);
        case 'edit':
            return renderWithClassName('diff-code-edit');
        default:
            return renderWithClassName(classNames(properties.className));
    }
};

const CodeCell = props => {
    const {text, tokens, renderToken, ...attributes} = props;
    const actualRenderToken = renderToken
        ? (token, i) => renderToken(token, defaultRenderToken, i)
        : defaultRenderToken;

    return (
        <td {...attributes}>
            {
                tokens
                    ? (tokens.length ? tokens.map(actualRenderToken) : ' ')
                    : (text || ' ')
            }
        </td>
    );
};

CodeCell.propTypes = {
    text: PropTypes.string.isRequired,
    tokens: PropTypes.arrayOf(PropTypes.object),
};

CodeCell.defaultProps = {
    tokens: null,
};

export default memo(CodeCell);
