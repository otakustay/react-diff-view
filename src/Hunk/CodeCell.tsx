import {HTMLAttributes, memo} from 'react';
import classNames from 'classnames';
import {TokenNode} from '../tokenize';
import {DefaultRenderToken, RenderToken} from '../context';

const defaultRenderToken: DefaultRenderToken = ({type, value, markType, properties, className, children}, i) => {
    const renderWithClassName = (className: string) => (
        <span key={i} className={className}>
            {value ? value : (children && children.map(defaultRenderToken))}
        </span>
    );


    switch (type) {
        case 'text':
            return value;
        case 'mark':
            return renderWithClassName(`diff-code-mark diff-code-mark-${markType}`);
        case 'edit':
            return renderWithClassName('diff-code-edit');
        default: {
            // properties normally not exist since it is deconstructed in pickRange, remove in next major release
            const legacyClassName = properties && properties.className;
            return renderWithClassName(classNames(className || legacyClassName));
        }
    }
};

function isEmptyToken(tokens: TokenNode[]) {
    if (!Array.isArray(tokens)) {
        return true;
    }

    if (tokens.length > 1) {
        return false;
    }

    if (tokens.length === 1) {
        const [token] = tokens;
        return token.type === 'text' && !token.value;
    }

    return true;
}

export interface CodeCellProps extends HTMLAttributes<HTMLTableCellElement> {
    changeKey: string;
    text: string;
    tokens: TokenNode[] | null;
    renderToken: RenderToken | undefined;
}

function CodeCell(props: CodeCellProps) {
    const {changeKey, text, tokens, renderToken, ...attributes} = props;
    const actualRenderToken: DefaultRenderToken = renderToken
        ? (token, i) => renderToken(token, defaultRenderToken, i)
        : defaultRenderToken;

    return (
        <td {...attributes} data-change-key={changeKey}>
            {
                tokens
                    ? (isEmptyToken(tokens) ? ' ' : tokens.map(actualRenderToken))
                    : (text || ' ')
            }
        </td>
    );
}

export default memo(CodeCell);
