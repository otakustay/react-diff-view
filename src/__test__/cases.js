/* In case you are looking for type of hunk, see src/Hunk/index.js */
import {parseDiff} from '..';

export const basic = `diff --git a/src/__test__/index.test.jsx b/src/__test__/index.test.jsx
index 643c2f0..7883597 100644
--- a/src/__test__/index.test.jsx
+++ b/src/__test__/index.test.jsx
@@ -21,3 +21,3 @@ describe('basic test', () => {
     test('App renders correctly', () => {
-        expect(renderer.create(<App diffText={'deff'} />).toJSON()).toMatchSnapshot();
+        expect(renderer.create(<App diffText={'diff'} />).toJSON()).toMatchSnapshot();
     });`;

export const basicHunk = parseDiff(basic)[0].hunks[0];

export const multiple = `diff --git a/src/index1.js b/src/index1.js
index e69de29..d00491f 100644
--- a/src/index1.js
+++ b/src/index1.js
@@ -0,0 +1 @@
+1
diff --git a/src/index.js b/src/index2.js
index d00491f..0cfbf08 100644
--- a/src/index2.js
+++ b/src/index2.js
@@ -1 +1 @@
-1
+2
diff --git a/src/index3.js b/src/index3.js
index 0cfbf08..e69de29 100644
--- a/src/index3.js
+++ b/src/index3.js
@@ -1 +0,0 @@
-2`;

export const multipleHunk = parseDiff(basic)[0].hunks[0];
