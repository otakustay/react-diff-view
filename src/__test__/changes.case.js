
export const normalChange = {isNormal: true, lineNumber: 0, oldLineNumber: 0};
export const insertChange = {isInsert: true, lineNumber: 0, oldLineNumber: 0};
export const deleteChange = {isDelete: true, lineNumber: 0, oldLineNumber: 0};
