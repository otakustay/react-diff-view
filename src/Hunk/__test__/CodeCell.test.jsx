import renderer from 'react-test-renderer';
import {basicHunk} from '../../__test__/cases';
import CodeCell from '../CodeCell';

const renderToken = (token, defaultRender, i) => {
    if (token.type === 'searchResult') {
        return (
            <span key={i} className="search-result">
                {token.children && token.children.map((token, i) => renderToken(token, defaultRender, i))}
            </span>
        );
    }
    return defaultRender(token, i);
};


describe('renderToken', () => {
    test('renders correctly', () => {
        expect(renderer.create(
            <CodeCell text="A" />
        ).toJSON()).toMatchSnapshot();
    });

    test('use renderToken', () => {
        expect(renderer.create(
            <CodeCell text="A" renderToken={renderToken} />
        ).toJSON()).toMatchSnapshot();
    });
});
