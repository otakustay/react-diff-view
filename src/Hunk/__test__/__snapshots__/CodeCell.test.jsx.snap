// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderToken renders correctly 1`] = `
<td>
  A
</td>
`;

exports[`renderToken use renderToken 1`] = `
<td>
  A
</td>
`;
